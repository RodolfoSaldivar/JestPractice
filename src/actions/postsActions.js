import axios from 'axios';
import {
	LOADING,
	ERROR,
	RELOAD,
	COM_LOADING,
	COM_ERROR,
	COM_RELOAD
} from '../types/postsTypes';
import * as usersTypes from '../types/usersTypes';

const { GET_ALL: USERS_GET_ALL } = usersTypes;

export const getByUser = (key) => async (dispatch, getState) => {
	dispatch({
		type: LOADING
	});

	let { users } = getState().usersReducer;
	const { posts } = getState().postsReducer;
	const user_id = users[key].id;

	try {
		const response = await axios.get(
			`https://jsonplaceholder.typicode.com/posts?userId=${user_id}`
		);
		const new_posts = response.data.map((post) => ({
			...post,
			comments: [],
			open: false
		}));
		const updated_posts = [...posts, new_posts];

		dispatch({
			type: RELOAD,
			payload: updated_posts
		});

		const posts_key = updated_posts.length - 1;
		const updated_users = [...users];
		updated_users[key] = {
			...users[key],
			posts_key
		};

		dispatch({
			type: USERS_GET_ALL,
			payload: updated_users
		});
	} catch (error) {
		console.log(error.message);
		dispatch({
			type: ERROR,
			payload: 'Posts unavailable.'
		});
	}
};

export const openClose = (pst_key, com_key) => (dispatch, getState) => {
	const { posts } = getState().postsReducer;
	const selected = posts[pst_key][com_key];

	const updated = {
		...selected,
		open: !selected.open
	};

	const updated_posts = [...posts];

	updated_posts[pst_key] = [...posts[pst_key]];
	updated_posts[pst_key][com_key] = updated;

	dispatch({
		type: RELOAD,
		payload: updated_posts
	});
};

export const getComments = (pst_key, com_key) => async (dispatch, getState) => {
	dispatch({
		type: COM_LOADING
	});

	const { posts } = getState().postsReducer;
	const selected = posts[pst_key][com_key];

	try {
		const response = await axios.get(
			`https://jsonplaceholder.typicode.com/comments?postId=${selected.id}`
		);

		const updated = {
			...selected,
			comments: response.data
		};

		const updated_posts = [...posts];

		updated_posts[pst_key] = [...posts[pst_key]];
		updated_posts[pst_key][com_key] = updated;

		dispatch({
			type: COM_RELOAD,
			payload: updated_posts
		});
	} catch (error) {
		console.log(error.message);
		dispatch({
			type: COM_ERROR,
			payload: 'Comments unavailable.'
		});
	}
};
