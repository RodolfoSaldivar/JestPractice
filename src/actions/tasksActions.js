import axios from 'axios';
import {
	GET_ALL,
	LOADING,
	ERROR,
	CHANGE_USER,
	CHANGE_TITLE,
	SAVED,
	RELOAD,
	CLEAR
} from '../types/tasksTypes';

export const getAll = () => async (dispatch) => {
	dispatch({
		type: LOADING
	});

	try {
		const response = await axios.get(
			'https://jsonplaceholder.typicode.com/todos'
		);

		const tasks = {};
		response.data.map(
			(tsk) =>
				(tasks[tsk.userId] = {
					...tasks[tsk.userId],
					[tsk.id]: {
						...tsk
					}
				})
		);

		dispatch({
			type: GET_ALL,
			payload: tasks
		});
	} catch (error) {
		console.log(error.message);
		dispatch({
			type: ERROR,
			payload: 'Tasks unavailable.'
		});
	}
};

export const changeUserId = (value) => (dispatch) => {
	dispatch({
		type: CHANGE_USER,
		payload: value
	});
};

export const changeTitle = (value) => (dispatch) => {
	dispatch({
		type: CHANGE_TITLE,
		payload: value
	});
};

export const add = (new_task) => async (dispatch) => {
	dispatch({
		type: LOADING
	});

	try {
		await axios.post('https://jsonplaceholder.typicode.com/todos', new_task);
		dispatch({
			type: SAVED
		});
	} catch (error) {
		console.log(error.message);
		dispatch({
			type: ERROR,
			payload: 'Try later.'
		});
	}
};

export const update = (updated_task) => async (dispatch) => {
	dispatch({
		type: LOADING
	});

	try {
		await axios.put(
			`https://jsonplaceholder.typicode.com/todos/${updated_task.id}`,
			updated_task
		);
		dispatch({
			type: SAVED
		});
	} catch (error) {
		console.log(error.message);
		dispatch({
			type: ERROR,
			payload: 'Try later.'
		});
	}
};

export const checkChange = (us_id, tsk_id) => (dispatch, getState) => {
	const { tasks } = getState().tasksReducer;
	const selected = tasks[us_id][tsk_id];

	const updated = {
		...tasks
	};
	updated[us_id] = {
		...tasks[us_id]
	};
	updated[us_id][tsk_id] = {
		...tasks[us_id][tsk_id],
		completed: !selected.completed
	};

	dispatch({
		type: RELOAD,
		payload: updated
	});
};

export const remove = (tsk_id) => async (dispatch) => {
	dispatch({
		type: LOADING
	});

	try {
		await axios.delete(
			`https://jsonplaceholder.typicode.com/todos/${tsk_id}`
		);
		dispatch({
			type: GET_ALL,
			payload: {}
		});
	} catch (error) {
		console.log(error.message);
		dispatch({
			type: ERROR,
			payload: 'Try later.'
		});
	}
};

export const clearForm = () => (dispatch) => {
	dispatch({
		type: CLEAR
	});
};
