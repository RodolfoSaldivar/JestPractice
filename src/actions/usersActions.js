import axios from 'axios';
import { GET_ALL, LOADING, ERROR } from '../types/usersTypes';

export const getAll = () => async (dispatch) => {
	dispatch({
		type: LOADING
	});

	console.log('inside user actions');

	try {
		const response = await axios.get(
			'https://jsonplaceholder.typicode.com/users'
		);
		dispatch({
			type: GET_ALL,
			payload: response.data
		});
	} catch (error) {
		console.log(error.message);
		dispatch({
			type: ERROR,
			payload: 'Try later.'
		});
	}
};
