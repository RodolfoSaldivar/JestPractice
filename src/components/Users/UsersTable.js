import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import Table from '@material-ui/core/Table';
import Paper from '@material-ui/core/Paper';
import Visibility from '@material-ui/icons/Visibility';
import TableRow from '@material-ui/core/TableRow';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import { makeStyles } from '@material-ui/core/styles';

//================================================

const useStyles = makeStyles((theme) => ({
	root: {
		width: '100%',
		marginTop: theme.spacing(3),
		overflowX: 'auto'
	},
	table: {
		minWidth: 650
	},
	icon: {
		color: theme.palette.text.primary
	}
}));

//================================================

export const UsersTable = (props) => {
	const classes = useStyles();

	const rows = () =>
		props.users.map((user, key) => (
			<TableRow key={user.id}>
				<TableCell>{user.name}</TableCell>
				<TableCell>{user.email}</TableCell>
				<TableCell>{user.website}</TableCell>
				<TableCell>
					<Link to={`/posts/${key}`}>
						<Visibility className={classes.icon} />
					</Link>
				</TableCell>
			</TableRow>
		));

	return (
		<div>
			<Paper className={classes.root} elevation={3}>
				<Table className={classes.table}>
					<TableHead>
						<TableRow>
							<TableCell>Name</TableCell>
							<TableCell>Mail</TableCell>
							<TableCell>Link</TableCell>
							<TableCell></TableCell>
						</TableRow>
					</TableHead>
					<TableBody>{rows()}</TableBody>
				</Table>
			</Paper>
		</div>
	);
};

const mapStateToProps = (reducers) => {
	return reducers.usersReducer;
};

export default connect(mapStateToProps)(UsersTable);
