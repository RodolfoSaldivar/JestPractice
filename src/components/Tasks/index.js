import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Spinner from '../General/Spinner';
import Fatal from '../General/Fatal';

import * as tasksActions from '../../actions/tasksActions';

class Tasks extends Component {
	componentDidMount() {
		if (!Object.keys(this.props.tasks).length) this.props.getAll();
	}

	componentDidUpdate() {
		const { tasks, loading, getAll } = this.props;

		if (!Object.keys(tasks).length && !loading) {
			getAll();
		}
	}

	content = () => {
		const { tasks, loading, error } = this.props;

		if (loading) {
			return <Spinner />;
		}
		if (error) {
			return <Fatal message={error} />;
		}

		return Object.keys(tasks).map((us_id) => (
			<div key={us_id}>
				<h2>User {us_id}</h2>
				<div className="contenedor_tareas">{this.ponerTasks(us_id)}</div>
			</div>
		));
	};

	ponerTasks = (us_id) => {
		const { tasks, checkChange, remove } = this.props;
		const per_user = {
			...tasks[us_id]
		};

		return Object.keys(per_user).map((tsk_id) => (
			<div key={tsk_id}>
				<input
					type="checkbox"
					defaultChecked={per_user[tsk_id].completed}
					onChange={() => checkChange(us_id, tsk_id)}
				/>
				{per_user[tsk_id].title}
				<button className="m_left">
					<Link to={`/tasks/save/${us_id}/${tsk_id}`}>Update</Link>
				</button>
				<button className="m_left" onClick={() => remove(tsk_id)}>
					Remove
				</button>
			</div>
		));
	};

	render() {
		return (
			<div>
				<button>
					<Link to="/tasks/save">Add</Link>
				</button>
				{this.content()}
			</div>
		);
	}
}

const mapStateToProps = ({ tasksReducer }) => tasksReducer;

export default connect(
	mapStateToProps,
	tasksActions
)(Tasks);
