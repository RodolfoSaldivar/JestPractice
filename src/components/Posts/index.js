import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';

import Spinner from '../General/Spinner';
import Fatal from '../General/Fatal';
import Comments from './Comments';

import * as usersActions from '../../actions/usersActions';
import * as postsActions from '../../actions/postsActions';

const { getAll: usersGetAll } = usersActions;
const { getByUser: postsGetByUser, openClose, getComments } = postsActions;

//================================================

const styles = (theme) => ({
	root: {
		width: '100%'
	}
});

//================================================

export class Posts extends Component {
	async componentDidMount() {
		const {
			usersGetAll,
			match: {
				params: { key }
			},
			postsGetByUser
		} = this.props;

		if (!this.props.usersReducer.users.length) {
			await usersGetAll();
		}
		if (this.props.usersReducer.error) {
			return;
		}
		if (!('posts_key' in this.props.usersReducer.users[key])) {
			await postsGetByUser(key);
		}
	}

	displayUser = () => {
		const {
			match: {
				params: { key }
			},
			usersReducer
		} = this.props;

		if (usersReducer.error) {
			return <Fatal message={usersReducer.error} />;
		}
		if (!usersReducer.users.length || usersReducer.loading) {
			return <Spinner />;
		}

		const name = usersReducer.users[key].name;

		return <h1>Posts of: {name}</h1>;
	};

	displayPosts = () => {
		const {
			usersReducer,
			usersReducer: { users },
			postsReducer,
			postsReducer: { posts },
			match: {
				params: { key }
			}
		} = this.props;

		if (!users.length) return;
		if (usersReducer.error) return;
		if (postsReducer.loading) {
			return <Spinner />;
		}
		if (postsReducer.error) {
			return <Fatal message={postsReducer.error} />;
		}
		if (!posts.length) return;
		if (!('posts_key' in users[key])) return;

		const { posts_key } = users[key];
		return this.showInfo(posts[posts_key], posts_key);
	};

	showInfo = (posts, post_key) =>
		posts.map((post, com_key) => (
			<ExpansionPanel
				key={post.id}
				onClick={() => this.showComments(post_key, com_key, post.comments)}
				elevation={3}
			>
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<h2>{post.title}</h2>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<div>
						<h3>{post.body}</h3>
						{post.open && <Comments comments={post.comments} />}
					</div>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		));

	showComments = (post_key, com_key, comments) => {
		this.props.openClose(post_key, com_key);
		if (!comments.length) {
			this.props.getComments(post_key, com_key);
		}
	};

	render() {
		const { classes } = this.props;
		return (
			<div>
				{this.displayUser()}
				<div className={classes.root}>{this.displayPosts()}</div>
			</div>
		);
	}
}

const mapStateToProps = ({ usersReducer, postsReducer }) => {
	return { usersReducer, postsReducer };
};

const mapDispatchToProps = {
	usersGetAll,
	postsGetByUser,
	openClose,
	getComments
};

export default withStyles(styles)(
	connect(
		mapStateToProps,
		mapDispatchToProps
	)(Posts)
);
