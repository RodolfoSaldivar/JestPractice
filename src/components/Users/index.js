import React, { Component } from 'react';
import { connect } from 'react-redux';
import Spinner from '../General/Spinner';
import Fatal from '../General/Fatal';
import UsersTable from './UsersTable';

import * as usersActions from '../../actions/usersActions';

export class Users extends Component {
	componentDidMount() {
		if (!this.props.users.length) {
			console.log('this.props.users: ', this.props.users);
			this.props.getAll();
			console.log('after getAll()');
		}
	}

	content = () => {
		if (this.props.loading) {
			return <Spinner />;
		}

		if (this.props.error) {
			return <Fatal message={this.props.error} />;
		}

		return <UsersTable />;
	};

	render() {
		return (
			<div>
				<h1>Users</h1>
				{this.content()}
			</div>
		);
	}
}

const mapStateToProps = (reducers) => {
	return reducers.usersReducer;
};

export default connect(
	mapStateToProps,
	usersActions
)(Users);
