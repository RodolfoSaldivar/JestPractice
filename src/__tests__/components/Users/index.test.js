import React from 'react';
import { shallow } from 'enzyme';
import fixtures from '../../fixtures/users';
import { Users } from '../../../components/Users';
import UsersTable from '../../../components/Users/UsersTable';
import * as usersActions from '../../../actions/usersActions';

const PROPS = {
	...fixtures.initial_state,
	...usersActions
};

describe('<Users />', () => {
	it('Renders correctly', () => {
		const users = shallow(<Users {...PROPS} />);
		expect(users).toMatchSnapshot();
	});

	it('Displays <UsersTable />', () => {
		const users = shallow(<Users {...PROPS} users={fixtures.one_user} />);
		expect(users).toMatchSnapshot();
		expect(users.find(UsersTable).exists()).toBe(true);
	});

	it('Displays <Spinner />', () => {
		const users = shallow(<Users {...PROPS} loading />);
		expect(users).toMatchSnapshot();
		expect(users.find('Spinner').exists()).toBe(true);
	});

	it('Displays <Fatal />', () => {
		const users = shallow(<Users {...PROPS} error="error" />);
		expect(users).toMatchSnapshot();
		expect(users.find('Fatal').exists()).toBe(true);
	});
});
