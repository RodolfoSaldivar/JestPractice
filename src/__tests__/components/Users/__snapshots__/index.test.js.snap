// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users /> Displays <Fatal /> 1`] = `
<div>
  <h1>
    Users
  </h1>
  <Fatal
    message="error"
  />
</div>
`;

exports[`<Users /> Displays <Spinner /> 1`] = `
<div>
  <h1>
    Users
  </h1>
  <Spinner />
</div>
`;

exports[`<Users /> Displays <UsersTable /> 1`] = `
<div>
  <h1>
    Users
  </h1>
  <Connect(UsersTable) />
</div>
`;

exports[`<Users /> Renders correctly 1`] = `
<div>
  <h1>
    Users
  </h1>
  <Connect(UsersTable) />
</div>
`;
